### PR TITLE
Bug 1529340 - CFR Pinned tabs popup animation

### PR DESCRIPTION
**Pull request by rlr**
This replaces pull #4827. This one is against the `cfr-pinned-tabs` branch.

The m-c changes that go with this are here: https://phabricator.services.mozilla.com/D22398

### CODE REVIEW

File: lib/CFRMessageProvider.jsm [hunk 1]
```diff
@@ -305,7 +305,6 @@ const CFR_MESSAGES = [
   {
     id: "PIN_TAB",
     template: "cfr_doorhanger",
-    exclude: true,
```
Comment (rlr) [inline lib/CFRMessageProvider.jsm]: do we want to get this "unexcluded" with this patch or should I undo this change?

File: lib/CFRPageActions.jsm [hunk 2]
```diff
@@ -275,6 +276,52 @@ class PageAction {
...
+    const footer = this.window.document.getElementById("cfr-notification-footer");
+    const ANIMATION_CONTAINER_ID = "cfr-notification-footer-pintab-animation-container";
+    let animationsEnabled = Services.prefs.getBoolPref(ANIMATIONS_ENABLED_PREF, true);
+    let animationContainer = this.window.document.getElementById(ANIMATION_CONTAINER_ID);
+    if (animationContainer) {
+      animationContainer.toggleAttribute("animate", animationsEnabled);
+      animationContainer.removeAttribute("paused");
+    } else {
+      animationContainer = this.window.document.createElement("vbox");
+      animationContainer.setAttribute("id", ANIMATION_CONTAINER_ID);
```
Comment (rlr) [inline lib/CFRPageActions.jsm]: I see a pattern of `.createElement(); .setAttribute("id",....)` that I guess could be factored out?

Comment (rlr) [reply lib/CFRPageActions.jsm]: along with the `.appendChild()` perhaps. 🙄

Comment (rlr) [reply lib/CFRPageActions.jsm]: something like this ?

```javascript
  _createElement(type, parent, id) {
    let element = this.window.document.createElement(type);
    if (id) {
      element.setAttribute("id", id);
    }
    parent.appendChild(element);
    return element;
  }

  async _renderPinTabAnimation() {
    const footer = this.window.document.getElementById("cfr-notification-footer");
    const ANIMATION_CONTAINER_ID = "cfr-notification-footer-pintab-animation-container";
    let animationsEnabled = Services.prefs.getBoolPref(ANIMATIONS_ENABLED_PREF, true);
    let animationContainer = this.window.document.getElementById(ANIMATION_CONTAINER_ID);
    if (animationContainer) {
      animationContainer.toggleAttribute("animate", animationsEnabled);
      animationContainer.removeAttribute("paused");
    } else {
      animationContainer = this._createElement("vbox", footer, ANIMATION_CONTAINER_ID);
      animationContainer.toggleAttribute("animate", animationsEnabled);

      // spacer
      this._createElement("vbox", animationContainer).setAttribute("flex", 1);

      let controlsContainer = this._createElement("hbox", animationContainer, "cfr-notification-footer-animation-controls");

      // spacer
      this._createElement("vbox", controlsContainer).setAttribute("flex", 1);

      let pauseButton = this._createElement("hbox", controlsContainer, "cfr-notification-footer-pause-button");
      pauseButton.addEventListener("click", () => {
        animationContainer.setAttribute("paused", true);
      });

      let pauseLabel = this._createElement("label", pauseButton, "cfr-notification-footer-pause-label");
      pauseLabel.textContent = await this.getStrings({"string_id": "cfr-doorhanger-pintab-animation-pause"});

      // pause icon
      this._createElement("image", pauseButton, "cfr-notification-footer-pause-icon");
    }
  }
```

**Comment by rlr** [comment]
The task-cluster failures are due to unreferenced images. That should fix itself once https://phabricator.services.mozilla.com/D22398 makes its way to m-c.

**Comment by rlr** [comment]
I ended up pushing the images in the m-c patch to avoid test failures on both sides.

File: lib/CFRPageActions.jsm (hunk 2 above)
Comment (piatra) [reply lib/CFRPageActions.jsm]: I like it. Writing unit tests that mess with the DOM is weird, this should make it nicer.

File: lib/CFRMessageProvider.jsm (hunk 1 above)
Comment (piatra) [reply lib/CFRMessageProvider.jsm]: undo it, we'll turn it on with a separate PR.

File: lib/CFRPageActions.jsm [hunk 3]
```diff
@@ -275,6 +276,52 @@ class PageAction {
...
+
+      let spacer = this.window.document.createElement("vbox");
+      spacer.setAttribute("flex", 1);
+      animationContainer.appendChild(spacer);
+
+      let controlsContainer = this.window.document.createElement("hbox");
+      controlsContainer.setAttribute("id", "cfr-notification-footer-animation-controls");
+
+      spacer = this.window.document.createElement("vbox");
+      spacer.setAttribute("flex", 1);
```
Comment (piatra) [inline lib/CFRPageActions.jsm]: Could this be replaced by `margins` in the m-c CSS?

File: lib/CFRPageActions.jsm (hunk 2 above)
Comment (piatra) [reply lib/CFRPageActions.jsm]: I would group `type` and `id` into an options object and have something like
```
_createElementAndAppend({id, type}, parent)
```

File: lib/CFRPageActions.jsm [hunk 4]
```diff
@@ -275,6 +276,52 @@ class PageAction {
...
+      controlsContainer.appendChild(spacer);
+
+      let pauseButton = this.window.document.createElement("hbox");
+      pauseButton.setAttribute("id", "cfr-notification-footer-pause-button");
+      pauseButton.addEventListener("click", () => {
```
Comment (piatra) [inline lib/CFRPageActions.jsm]: We need to remove this event listener.

Comment (piatra) [reply lib/CFRPageActions.jsm]: Probably in `_popupStateChange`

File: lib/CFRPageActions.jsm [hunk 5]
```diff
@@ -275,6 +276,52 @@ class PageAction {
...
+      });
+
+      let pauseLabel = this.window.document.createElement("label");
+      pauseLabel.setAttribute("id", "cfr-notification-footer-pause-label");
+      pauseLabel.textContent = await this.getStrings({"string_id": "cfr-doorhanger-pintab-animation-pause"});
+      pauseButton.appendChild(pauseLabel);
+
+      let pauseIcon = this.window.document.createElement("image");
+      pauseIcon.setAttribute("id", "cfr-notification-footer-pause-icon");
+      pauseButton.appendChild(pauseIcon);
```
Comment (piatra) [inline lib/CFRPageActions.jsm]: Could we add the icon via CSS as background so we don't create so much markup?

File: lib/CFRPageActions.jsm (hunk 3 above)
Comment (rlr) [reply lib/CFRPageActions.jsm]: possibly? Would it auto-grow? But the m-c change is already in autoland so it would have to be a followup.

File: lib/CFRPageActions.jsm (hunk 5 above)
Comment (rlr) [reply lib/CFRPageActions.jsm]: it is CSS but I guess maybe it didnt need it's own element. also, a followup since m-c change already shipped.

**Comment by rlr** [comment]
The images also shipped with the m-c CSS changes so tests would pass (unreferenced file failures). So I guess I have a list of things I can clean up next week. I can file a followup.

File: lib/CFRPageActions.jsm (hunk 4 above)
Comment (rlr) [reply lib/CFRPageActions.jsm]: I just added it on `window.unload`.

File: lib/CFRPageActions.jsm (hunk 3 above)
Comment (rlr) [reply lib/CFRPageActions.jsm]: Oh I see. You mean not using a spacer and just setting padding-top or margin-top on the container. Makes sense 👍

**Comment by rlr** [comment]
I'll file followup bugs to:
* simplify the markup:
** remove some unnecessary elements for spacing and the icon and update the CSS accordingly
** get and use 2X res images
** make the images configurable (this would be for Firefox 68+, we don't need it in 67)

Anything else?

**Comment by rlr** [comment]
Taskcluster should be green once https://hg.mozilla.org/integration/autoland/rev/6a3b22130b996d2ea931525093f3e6103ed13d20 makes its way to m-c

File: lib/CFRPageActions.jsm (hunk 4 above)
Comment (k88hudson) [reply lib/CFRPageActions.jsm]: I don't think this is strictly necessary as-is considering the event listener gets added only on element creation and presumably the element gets destroyed on `window.unload` anyway, although I wonder if it would be more appropriate to attach/remove the event listener on `_expand`/`_collapse` so that it's not long-lived throughout the whole life of the window. What do you think?

File: test/unit/asrouter/CFRPageActions.test.js [hunk 6]
```diff
@@ -110,6 +113,8 @@ describe("CFRPageActions", () => {
     for (const id of elementIDs) {
       const elem = document.createElement("div");
       elem.setAttribute("id", id);
+      // TODO: Remove this once travis is on Firefox 63+
+      elem.toggleAttribute = () => {};
```
Comment (k88hudson) [inline test/unit/asrouter/CFRPageActions.test.js]: lol

File: lib/CFRPageActions.jsm [hunk 7]
```diff
@@ -275,6 +284,47 @@ class PageAction {
...
+
+      // pause icon
+      this._createElement({type: "image", id: "cfr-notification-footer-pause-icon"}, pauseButton);
+    }
+
+    animationContainer.toggleAttribute("animate", Services.prefs.getBoolPref(ANIMATIONS_ENABLED_PREF, true));
+    animationContainer.removeAttribute("paused");
+
+    let pauseButton = this.window.document.getElementById(PAUSE_BUTTON_ID);
+    this.onPauseClick = () => { animationContainer.setAttribute("paused", true); };
```
Comment (k88hudson) [inline lib/CFRPageActions.jsm]: you could also put these lines under `if !(this.onPauseClick) { }` just in case

Comment (rlr) [reply lib/CFRPageActions.jsm]: yep, that's a good idea 👍

**Comment by k88hudson** [comment]
Ok, I'm merging this to `mozilla/cfr-pinned-tabs` and rebasing on the export.
